### PR TITLE
For review: memory efficient KL loss

### PR DESCRIPTION
**Pull request by justheuristic**

### CODE REVIEW

File: src/memory_efficient_loss.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    for chunk_start in range(0, total_tokens, max_tokens_per_chunk):
+        is_last_chunk = chunk_start + max_tokens_per_chunk >= total_tokens
+        loss_values_by_chunk.append(maybe_checkpoint(
+            _compute_kl_div_from_flat_hidden_states,
+            flat_student_hidden_states[chunk_start: chunk_start + max_tokens_per_chunk], student_lm_head,
+            flat_teacher_hidden_states[chunk_start: chunk_start + max_tokens_per_chunk], teacher_lm_head,
+            checkpoint_enabled=torch.is_grad_enabled() and (checkpoint_last_chunk or not is_last_chunk),
+            **checkpoint_kwargs
+        ))
+    return torch.cat(loss_values_by_chunk).reshape(*student_hidden_states.shape[:2])
```
Comment (galqiwi) [inline src/memory_efficient_loss.py:50]: Newline before return? (optional)

File: src/memory_efficient_loss.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    teacher_hidden_states = torch.randn(batch_size, seq_length, teacher_hidden_size)
+    student_hidden_states = torch.randn(batch_size, seq_length, student_hidden_size, requires_grad=True)
+
+    ref_loss_values = F.kl_div(
+        input=F.log_softmax(student_lm_head(student_hidden_states), dim=-1),
+        target=F.log_softmax(teacher_lm_head(teacher_hidden_states), dim=-1),
+        log_target=True, reduction="none"
+    ).sum(-1)
+
+    for use_reentrant in [True, False]:
```
Comment (galqiwi) [inline src/memory_efficient_loss.py:89]: Let's do
```
for use_reentrant, checkpoint_last_chunk, determinism_check in itertools.product(
    (True, False),
    (True, False),
    ("default", "none"),
):
```
instead.

https://github.com/torvalds/linux/blob/master/Documentation/process/coding-style.rst?plain=1#L33

**Review by galqiwi — COMMENTED** [review]
You can also fix isort if you want to, but we'll anyway do it automatically later, so I don't care

File: src/memory_efficient_loss.py (hunk 1 above)
Comment (justheuristic) [reply src/memory_efficient_loss.py:51]: [agreed to keep as is in a 1on1 chat]

File: src/memory_efficient_loss.py (hunk 2 above)
Comment (justheuristic) [reply src/memory_efficient_loss.py]: Updated it, thanks!
